### DISCUSSION TITLE
How to avoid skipReason invalid-name in the context of Docker build args

### DISCUSSION THREAD

**Discussion by reitzig** [renovatebot/renovate#24253] [Request Help] [answered]
### How are you running Renovate?

Self-hosted

### If you're self-hosting Renovate, tell us what version of Renovate you run.

36.81.0

### If you're self-hosting Renovate, select which platform you are using.

Bitbucket Server

### Was this something which used to work for you, and then stopped?

I never saw this working

### Wanted end result.

I want to avoid "warnings" of this type:

```json
{
   "skipReason": "invalid-name",
   "replaceString": "$base_image",
   "autoReplaceStringTemplate": "{{depName}}{{#if newValue}}:{{newValue}}{{/if}}{{#if newDigest}}@{{newDigest}}{{/if}}",
   "datasource": "docker",
   "depType": "stage",
   "updates": []
 }
```

This comes from something like

```dockerfile
ARG base_image
FROM $base_image
```

This will never be a Renovate-able dependency (in `Dockerfile`, that is; I _do_ Renovate it build-site!) so I want to suppress it.

### What you tried so far.

```json
{
    "packageRules": [
          {
            "matchFileNames": ["**/Dockerfile*"],
            "matchDatasources" : ["docker"],
            "matchPackageNames" : ["$base_image"],
            "enabled": false
          }
    ]
}
```

This was not effective. By my guess, `invalid-name` triggers _before_ (or during) package/dep name extraction so the rules doesn't apply. I don't see `matchReplaceString` in the documentation, or similar.

**Comment by rarkins** [comment]
How is it a "warning"?

**Reply by reitzig** [reply]
FWIW, context: [renovatebot/renovate#23961.1697777181](https://github.com/renovatebot/renovate/issues/23961#issuecomment-1697777181) with [renovatebot/renovate#24156](https://github.com/renovatebot/renovate/issues/24156) in mind.

There are several `skipReason`s that (can) hint at problems that the developers should address (IMHO). In total, I'm seeing these in our logs:

```
(I)  disabled
(I)  file
(I)  internal-package
(W)  invalid-name
(W)  invalid-value
(W)  invalid-version
(I)  local
(I)  local-dependency
(I)  name-placeholder
(W)  unspecified-version
(W)  unsupported
(I)  version-placeholder
```

Based on some spot checks, I've classified them as ignore (I) and warn (W), with the idea that devs should investigate any (W)s.

Now, in _this_ case, I am the developer, and think the invalid name is okay and should be ignored. 
At the same time, I (Renovate-log parsing me) am mostly guessing when which skip reason is used, but I don't think that's a general rule?

Now that I look at that list again: shouldn't this case be `name-placeholder`? 🤔

**Reply by rarkins** [reply]
Thanks, I understand what you mean now. Skip reasons aren't meant as a _warning_ but you're welcome to submit a PR which gives it a more specific skip reason in such cases

**Reply by reitzig** [reply]
>  Skip reasons aren't meant as a warning

If I may ask: What _are_ they meant for?

While debugging custom (regex) managers, they were often the only indication something wasn't working correctly (and how so, however brief the feedback). Or so I thought; did I miss something? (ah, throwback to #20881 on which I owe you responses -- sorry! 🙈)

> submit a PR which gives it a more specific skip reason in such cases

I don't see where `invalid-name` is emitted in [renovatebot/renovate:lib/modules/manager/dockerfile/extract.ts](https://github.com/renovatebot/renovate/blob/main/lib/modules/manager/dockerfile/extract.ts). 🤔 Is [renovatebot/renovate:lib/workers/repository/process/fetch.ts#46](https://github.com/renovatebot/renovate/blob/667f1370819d5bb128459fdf7b2c232d01b0ef6f/lib/workers/repository/process/fetch.ts#L46) responsible? In that case, it seems the default value for the ARG variable (empty, here) is being picked up by Renovate, which is only sometimes helpful. Anyway, seems like more logic needs investigating than I can spare time for, sorry. 😕

That said, it presents the opportunity for a dirty workaround: put down any dummy value as default for the ARG and disable based on _that_ value.

**Reply by rarkins** [reply, marked as answer]
`skipReason` is an internal concept meant to provide some category of why a dependency is skipped and not updated

**Comment by reitzig** [comment]
Workaround confirmed:

```dockerfile
ARG base_image="dummy-for-renovate"
FROM $base_image
```

together with

```json5
{
  "packageRules" : [
    {
      "description": "Disable dependencies that are managed by build arg, not in Dockerfile \
                      (identified by dummy name)",
      "matchFileNames": ["**/Dockerfile*"],
      "matchDatasources" : ["docker"],
      "matchPackageNames" : ["dummy-for-renovate"],
      "enabled": false
    }
  ]
}
```

Results in:

```json
{
  "depName":                   "dummy-for-renovate",
  "replaceString":             "dummy-for-renovate",
  "autoReplaceStringTemplate": "{{depName}}{{#if newValue}}:{{newValue}}{{/if}}{{#if newDigest}}@{{newDigest}}{{/if}}",
  "datasource":                "docker",
  "depType":                   "stage",
  "updates":                   [],
  "packageName":               "dummy-for-renovate",
  "skipReason":                "disabled"
}
```

Note `disabled` instead of `invalid-name`, which arguably matches intent besser. Is it worth it? Depdends, I guess. 🤷
